src/motion/features/viewport/observers.ts

```diff
@@ -0,0 +1,71 @@
+type IntersectionHandler = (entry: IntersectionObserverEntry) => void
+
+interface ElementIntersectionObservers {
+    [key: string]: IntersectionObserver
+}
+
+/**
+ * Map an IntersectionHandler callback to an element. We only ever make one handler for one
+ * element, so even though these handlers might all be triggered by different
+ * observers, we can keep them in the same map.
+ */
+const observerCallbacks = new WeakMap<Element, IntersectionHandler>()
+
+/**
+ * Multiple observers can be created for multiple element/document roots. Each with
+ * different settings. So here we store dictionaries of observers to each root,
+ * using serialised settings (threshold/margin) as lookup keys.
+ */
+const observers = new WeakMap<
+    Element | Document,
+    ElementIntersectionObservers
+>()
+
+const fireObserverCallback = (entry: IntersectionObserverEntry) => {
+    observerCallbacks.get(entry.target)?.(entry)
+}
+
+const fireAllObserverCallbacks: IntersectionObserverCallback = (entries) => {
+    entries.forEach(fireObserverCallback)
+}
+
+function initIntersectionObserver({
+    root,
+    ...options
+}: IntersectionObserverInit): IntersectionObserver {
+    /**
+     * If we don't have an observer lookup map for this root, create one.
+     */
+    if (!observers.has(root || document)) {
+        observers.set(root || document, {})
+    }
+    const rootObservers = observers.get(root || document)!
+
+    const key = JSON.stringify(options)
+
+    /**
+     * If we don't have an observer for this combination of root and settings,
+     * create one.
+     */
+    if (!rootObservers[key]) {
+        rootObservers[key] = new IntersectionObserver(
+            fireAllObserverCallbacks,
+            { root, ...options }
+        )
+    }
+
+    return rootObservers[key]
+}
+
+export function createIntersectionObserver(
+    element: Element,
+    options: IntersectionObserverInit,
+    callback: IntersectionHandler
+) {
+    const rootInteresectionObserver = initIntersectionObserver(options)
+
+    observerCallbacks.set(element, callback)
+    rootInteresectionObserver.observe(element)
+
+    return () => rootInteresectionObserver.unobserve(element)
```

Comment:
I would say that this shouldn't be part of creating the observer, and it leads to a bit weird code where we return the result from this in useEffect. I don't feel strongly, but shouldn't we pull this out?

Reply:
We wrapped the IntersectionObserver in a separate class for this [here](https://github.com/framer/FramerStudio/blob/25d3c81abbc4638e84e371cc9c62883b6c72142c/src/library/src/modules/withVariantAppearEffect.tsx#L5)

Reply:
What about:

```typescript
export function createIntersectionObserver(
    options: IntersectionObserverInit,
    callback: IntersectionHandler
) {
    const rootInteresectionObserver = initIntersectionObserver(options)

    const observe = (element: Element) => {
        observerCallbacks.set(element, callback)
        rootInteresectionObserver.observe(element)
    }
    const unobserve = (element: Element) => {
        rootInteresectionObserver.unobserve(element)
    }

    return { observe, unobserve }
}
```

and then in `useIntersectionObserver` we can do:

```typescript
useEffect(() => {
    if (!shouldObserve) return

    const {observe, unobserve} = createIntersectionObserver(...)
    const element = visualElement.getInstance()
    observe(element)
    return () => unobserve(element)
})
```

Reply:
I'm a big believer in subscribers returning unsubscriptors and it's quite a common pattern in Motion. For instance `value.onChange`, `SubscriptionManager` and so on. I don't see the need for a wrapper class?

Reply:
Yeah I can see that and agree. Maybe it's more about the naming, I wouldn't expect a function that is called `create...Observer` to also start to observe something.
Ok, last try:

useEffect(() => {
    if (!shouldObserve) return
    const observe = createIntersectionObserver(...)
    return observe(visualElement.getInstance())
})

As I said, I don't feel strongly, so do what you think is best.

Reply:
And the wrapper class is not needed indeed, but I just wanted to point out that we had a similar problem in in the useAppearEffect hook

Reply:
Oh I see it's more about the observation. I think just renaming to `startObservingIntersection`, `observeIntersection` or similar is fine then, creating the intersection observer itself is in some ways an implementation detail.

We can delete the observer too although my suspicion was it's unneeded because by disconnecting the observer we remove it from `entries` and, if the element itself has been removed, so too will the reference in the WeakMap. But there's no problem being explicit here just in case.

Reply:
Ah, yeah, that rename sounds great!